src/DDMinimizer.cpp

```diff
@@ -0,0 +1,445 @@
+    int max = -1;
+    for (const auto& pair : map.second) {
+      if (pair.second == -1) {
+        max = -1;
+        break;
+      }
+      if (pair.second > max) {
+        max = pair.second;
+      }
+    }
+    std::string mapName = map.first.substr(0, 2);
+    auto it = indices.find(mapName);
+
+    if (it != indices.end()) {
+      if (mapName[1] == 'R') {
+        const auto column = static_cast<std::size_t>(map.first[2] - '0');
+        it->second[bits - 1 - column] = max;
+      } else if (mapName[1] == 'L') {
+        const auto column = static_cast<std::size_t>(map.first[2] - '0');
+        it->second[column] = max;
+      } else {
+        it->second[0] = max;
+      }
+    }
+  }
+
+  // create the permutation based on the order of max index in the complete maps
+  std::vector<Qubit> layout(bits);
+  std::iota(layout.begin(), layout.end(), 0);
+
+  auto cX = indices.find("cX");
+  auto xC = indices.find("xC");
+  auto cL = indices.find("cL");
+  auto xL = indices.find("xL");
+  auto cR = indices.find("cR");
+  auto xR = indices.find("xR");
+
+  const int prioCr = DDMinimizer::getLadderPosition(cR->second, xC->second[0]);
+  const int prioXl = DDMinimizer::getLadderPosition(xL->second, xC->second[0]);
+  const int stairsCr = DDMinimizer::getStairCount(cR->second);
+  const int stairsXl = DDMinimizer::getStairCount(xL->second);
+  const int prioCl = DDMinimizer::getLadderPosition(cL->second, cX->second[0]);
+  const int prioXr = DDMinimizer::getLadderPosition(xR->second, cX->second[0]);
+  const int stairsCl = DDMinimizer::getStairCount(cL->second);
+  const int stairsXr = DDMinimizer::getStairCount(xR->second);
+
+  // complete case checkins and adjust the layout
+  if ((cX->second[0] != -1) &&
+      (xC->second[0] == -1 || cX->second[0] < xC->second[0])) {
+
+    if (prioCr == 0 && stairsCr > 0) {
+      layout = DDMinimizer::reverseLayout(layout);
+      layout = DDMinimizer::rotateLeft(layout, stairsCr);
+    } else if (prioXl == 0 && stairsXl > 0) {
+      layout = DDMinimizer::reverseLayout(layout);
+      layout = DDMinimizer::rotateRight(layout, stairsXl);
+    } else if (prioCr > 0 || prioXl > 0 || (prioCr == 0 && prioXl == 0)) {
+      layout = DDMinimizer::reverseLayout(layout);
+    }
+  } else if ((xC->second[0] != -1) &&
+             (cX->second[0] == -1 || cX->second[0] > xC->second[0])) {
+
+    if ((prioCl == 0 && DDMinimizer::isFull(cL->second)) ||
+        (prioXr == 0 && DDMinimizer::isFull(xR->second))) {
+      layout = DDMinimizer::reverseLayout(layout);
+    }
+
+    else if (prioCl == 0 && stairsCl > 0) {
+      layout = DDMinimizer::rotateRight(layout, stairsCl);
+    } else if (prioXr == 0 && stairsXr > 0) {
+      layout = DDMinimizer::rotateLeft(layout, stairsXr);
+    }
+  } else if ((DDMinimizer::isFull(xR->second) ||
+              DDMinimizer::isFull(cL->second))) {
+    layout = DDMinimizer::reverseLayout(layout);
+  } else {
+    return DDMinimizer::createControlBasedPermutation(qc);
```

Comment:
This case seems fundamentally different from the other cases. Would you mind leaving a comment in the code on why this function is called?